src/lib/models/emulation/Core.cc

```diff
@@ -118,19 +97,23 @@ void Core::tick() {
       return;
     }
     if (addresses.size() > 0) {
-      // Memory reads are required; request them, set `pendingReads_`
-      // accordingly, and end the cycle early
+      // Memory reads required; request them
       for (auto const& target : addresses) {
         dataMemory_.requestRead(target);
-        // Store addresses for use by next store data operation
+        // Save addresses for use by instructions that perform a LD and STR
+        // (i.e. single instruction atomics)
         previousAddresses_.push_back(target);
       }
-      pendingReads_ = addresses.size();
-      return;
-    } else {
-      // Early execution due to lacking addresses
-      execute(uop);
-      return;
+      // Emulation core can only be used with a Flat memory interface, so data
+      // is ready immediately
```

Comment:
Is there a way that this can be enforced through an assert or different typing? I think currently the only line of defence is the checks in modelConfig:1040